parlai/core/teachers.py

```diff
@@ -775,11 +785,24 @@ def get(self, episode_idx, entry_idx=0):
             which example to return from the episode. Many datasets have only
             single-entry episodes, so this defaults to zero.
         """
+        next_episode_idx_for_rank = episode_idx + 1
+        if self.is_distributed_and_is_eval:
+            raw_episode_idx = episode_idx
+            episode_idx = raw_episode_idx * self.num_workers + self.rank
+            next_episode_idx_for_rank = episode_idx + self.num_workers
+
+            if episode_idx >= len(self.data):
+                # This can occur in spite of the check below if epoch ends
+                # mid-batch b/c BatchWorld calls act() on all the worlds without
+                # checking if epochDone
+                return {'episode_done': True}, True
+
         # first look up data
         episode = self.data[episode_idx]
         entry = episode[entry_idx]
         episode_done = entry_idx == len(episode) - 1
-        end_of_data = episode_done and episode_idx == len(self.data) - 1
+
+        end_of_data = episode_done and next_episode_idx_for_rank >= len(self.data)
```

Comment:
this looks like it will have an off by one compared to the line you deleted

Reply:
Don't think so? Can you take another look? I think the tests would fail too if I did, b/c it would try to get again and would try to do self.data[episode_idx] that doesn't exist.

Reply:
I see, because next_episode_idx_for_rank != episode_idx.

Reply:
Ya exactly